src/core/lib/layer-manager.js

```diff
@@ -203,32 +205,38 @@ export default class LayerManager {
 
   updateLayers({newLayers}) {
     // TODO - something is generating state updates that cause rerender of the same
-    if (newLayers === this.lastRenderedLayers) {
+    if (newLayers === this.lastRenderedLayers && !this._isContextDirty) {
       log.log(3, 'Ignoring layer update due to layer array not changed');
       return this;
     }
-    this.lastRenderedLayers = newLayers;
 
-    assert(this.context.viewport, 'LayerManager.updateLayers: viewport not set');
+    if (newLayers !== this.lastRenderedLayers) {
+      this.lastRenderedLayers = newLayers;
 
-    // Filter out any null layers
-    newLayers = newLayers.filter(newLayer => newLayer !== null);
+      assert(this.context.viewport, 'LayerManager.updateLayers: viewport not set');
 
-    for (const layer of newLayers) {
-      layer.context = this.context;
-    }
+      // Filter out any null layers
+      newLayers = newLayers.filter(newLayer => newLayer !== null);
 
-    this.prevLayers = this.layers;
-    const {error, generatedLayers} = this._updateLayers({
-      oldLayers: this.prevLayers,
-      newLayers
-    });
+      this.prevLayers = this.layers;
+      const {error, generatedLayers} = this._updateLayers({
+        oldLayers: this.prevLayers,
+        newLayers
+      });
 
-    this.layers = generatedLayers;
-    // Throw first error found, if any
-    if (error) {
-      throw error;
+      this.layers = generatedLayers;
+      // Throw first error found, if any
+      if (error) {
+        throw error;
+      }
     }
+
+    // Update layers context
+    for (const layer of this.layers) {
```

Comment:
Only change here is moving this for loop out as it applies to both cases (newLayers and context updated), and looping over this.layers instead of newLayers, as they shouldn't differ.

Reply:
I think `context` needs to be set during the _updateLayers cycle. That cycle calls layer life cycle callbacks and context must be set by then. So moving this out of the loop will at minimum do duplicate work.